pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,303 @@
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return false
+	}
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return false
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+	if clusterName != "admin" {
+		//	klog.V(2).Infof("Skipping update for non-admin cluster %q", clusterName)
+		//	return false
+	}
+
+	current, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.V(2).Infof("Object was not Unstructured: %T", obj)
+		return false
+	}
+
+	if namespaceBlocklist.Has(current.GetNamespace()) {
+		klog.V(2).Infof("Skipping syncing namespace %q", current.GetNamespace())
+		return false
+	}
+	return true
+}
+
+func filterNamespace(obj interface{}) bool {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return false
+	}
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return false
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+	if clusterName != "admin" {
+		//	klog.Infof("Skipping update for non-admin cluster %q", clusterName)
+		//	return false
+	}
+	if namespaceBlocklist.Has(name) {
+		klog.V(2).Infof("Skipping syncing namespace %q", name)
+		return false
+	}
+	return true
+}
+
+func (c *Controller) enqueueResource(gvr schema.GroupVersionResource, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	gvrstr := strings.Join([]string{gvr.Resource, gvr.Version, gvr.Group}, ".")
+	c.queue.AddRateLimited(gvrstr + "::" + key)
+}
+
+func (c *Controller) enqueueNamespace(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("namespace::" + key)
+}
+
+func (c *Controller) enqueueCluster(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("cluster::" + key)
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer c.queue.ShutDown()
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+	klog.Infof("Starting workers")
+	<-ctx.Done()
+	klog.Infof("Stopping workers")
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(ctx, strings.TrimPrefix(key, "namespace::"))
```

Comment:
Speaking for my interpretation of their comments on the Workspace PR, I think Andy and Stefan would have expected a queue to be for one type and one type only

Reply:
It's good advice. I don't think we'd be able to have a queue per-dynamic-resource-type, but we could definitely split out namespace and cluster items from arbitrary-resource updates. I'll do that now.